### PR TITLE
client-server connection

### PR DESCRIPTION
**Pull request by furncyn**
Estalish client-server connection and implement a simple communication over connection

### CODE REVIEW

File: Client.cpp [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+	cerr << "ERROR: " + s << endl;
+}
+
+void Client::init(string hostname, int port, string filename) {
+	// Create a UDP socket
+	if((this->socket_fd = socket(AF_INET, SOCK_DGRAM, 0)) < 0) {
+		error("socket creation failed");
+		exit(1);
+	}
+	bzero(&(this->servaddr), sizeof(this->servaddr));
```
Comment (GalenWong) [inline Client.cpp]: I don't know why... the example code online (especially GeeksForGeeks) always tells us to explicitly zero-initialize variables with `bzero` or some other method of filling in 0s. In fact, by the C++11 standard, all values initialized locally to a thread (like how we do it in `client.c`) are all zero-initialized, as explained [here](https://blog.tartanllama.xyz/initialization-is-bonkers/) and [here](https://en.cppreference.com/w/cpp/language/zero_initialization).

I am not saying explicitness like this is not good. It is just funny to think about if people are explicitly being explicit to be beginner friendly or they are just ignorant of the C++ standard. Or maybe they wrote everything in the older C++ standards. 

Anyway, this is just Galen's weird contemplation of codes that don't matter that much lol. You can decide to change the code or not.

File: Client.cpp [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+	// Resolving hostname
+	hostent* record = gethostbyname(hostname.c_str());
```
Comment (GalenWong) [inline Client.cpp]: `gethostbyname` is obsolete according to the [man page](http://man7.org/linux/man-pages/man3/gethostbyname.3.html#DESCRIPTION). Please use `getaddrinfo`. There is an example that sets up a UDP connection in the [man page](http://man7.org/linux/man-pages/man3/getaddrinfo.3.html#EXAMPLE) as well.

File: Client.cpp [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+	string message = "Hello Server";
+	if(sendto(this->socket_fd, message.c_str(), 1024, 0, (struct sockaddr*) NULL, sizeof(this->servaddr)) < 0) {
+		error("sendto failed");
+		cout << strerror(errno) << endl;
+	}
+	// waiting for response
+	char buffer[1024];
+	recvfrom(this->socket_fd, buffer, sizeof(buffer), 0, (struct sockaddr*) NULL, NULL);
+	puts(buffer);
+	close(this->socket_fd);
```
Comment (GalenWong) [inline Client.cpp]: We will reimplement this to be handshake procedure. But we should do it in a separate PR. Keep this for now.

File: Server.cpp [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+	len = sizeof(this->cliaddr);
+	int n = recvfrom(this->socket_fd, buffer, sizeof(buffer), 0, (struct sockaddr*) &(this->cliaddr), &len);
+	buffer[n] = '\0';
+	puts(buffer);
+	string message = "Hello Client";
+	if(sendto(this->socket_fd, message.c_str(), 1024, 0, (struct sockaddr*) &(this->cliaddr), sizeof(this->cliaddr)) < 0) {
+		error("sendto failed");
+		cerr << strerror(errno) << endl;
+	}
+}
```
Comment (GalenWong) [inline Server.cpp]: This will also be discarded in the next PR.

**Comment by furncyn** [comment]
@GalenWong @GalenWong @GalenWong
